### DISCUSSION TITLE
morphologist 2023 won't open: AttributeError: 'NoneType' object has no attribute '_parameterized'

### DISCUSSION THREAD

**Discussion by HdeV999** [brainvisa/brainvisa.github.io#180] [Q&A] [answered]
Dear BrainVISA team, 
I was happily running batch processing scripts when something unadvisable happened: I exceeded the quotas of my home directory during the processing, and it affected my installation. The most obvious result was that my .brainvisa/options.minf was overwritten with a blank file, but I believe I solved that by reconfiguring my options. 
A side effect is that I can't open the morphologist 2023 process anymore: I get a "AttributeError: 'NoneType' object has no attribute '_parameterized'" error when I try to open the morphologist 2023 window. 
Incidentally, when I try to run the magic command bv_maker, I fail the source step: master (1e06678e) x merge failed U-49  axon/master
This should be unrelated but may still be relevant to try and solve the issue...
This is my environment specification: 
opensource-master-5.3
  type: dev
  distro: opensource
  branch: master
  system: ubuntu-22.04
  image_version: 5.3
  container_type: singularity
  image: /home/hd488/casa_distro/casa-dev-5.3-22.sif
  directory: /home/hd488/casa_distro/brainvisa-opensource-master
Thanks for any help on the matter...

**Comment by HdeV999** [comment]
Update: I tried running a bv_maker configure build doc to restore my previous build even without being able to run the source step, it worked, but I'm still faced with the same morphologist 2023 issue, i.e. when I double click on the process I get an error window with 
AttributeError: 'NoneType' object has no attribute '_parameterized'

**Comment by denisri** [comment, marked as answer]
Perhaps the disk full issue happened during a git operation.
Try going to src/axon/master
and there, do:
```
git reset --hard origin/master
git pull
```
then re-run bv_maker

**Reply by HdeV999** [reply]
Thanks! It most probably works... I did something which probably accounts to the same but in a much more bruteforce way: I started recreating a new brainvisa from the same .sif file, and just copy pasted the new src/axon folder to my initial brainvisa install, then deleted the "new install" and ran bv_maker. It has resolved the issue. I'll remember the clean way for the future!

**Reply by HdeV999** [reply]
I confirm that it works... Managed to re-fill my homedir... Could you please remind how to disable the option to automatically generate a visualization of the automated labelling after running the morphologist pipeline please?

**Reply by denisri** [reply]
Hi,
You're talking about the new "report" process at the end of the pipeline ? You can disable the step. If you are running it using axon-runprocess, you can specify it as a parameter of the pipeline as:
```
axon-runprocess morphologist Report.selected=False ...
```
Denis

**Reply by HdeV999** [reply]
Wonderful thank you. I'm using brainvisa.axon.runprocess in a workflow, and actually using Import_FROM_FreeSurfer_TO_Morpho and not directly morphologist. 
The command I use to create my workflow is this one: 
command_fstomorpho = base_cmd + options_fstomorpho + write_wf_fstomorpho + process_fstomorpho + parameters_fstomorpho with:
base_cmd = ["axon-runprocess"]
options_fstomorpho = ["-i", "T1_orig", ...]
wite_wf_fstomorpho = ["-w", "path/to/my/workflow"]
process_fstomorpho = ["Import_FROM_FreeSurfer_TO_Morpho"]
parameters_fstomorpho = ["T1_orig=" + repr(orig), ...]

Can I just add as a first option ["-i", "Report.selected"] and as a first parameter ["Report.selected=" + repr(False)]?

**Reply by c-langlet** [reply]
I think you dont need to iterate on this parameter, if you want it applied to all iterations you just have to add  ["Report.selected=False"]

**Reply by HdeV999** [reply]
As a parameter and not specified in options then?

**Reply by denisri** [reply]
Uh oh it's not directly the morphologist pipeline, but the freesurfer import process, so what I told you won't work: here the morphologist pipeline is used internally and the parameters are not exposed. A modification of the process will be needed then.
Does it display things on screen, even when it is used in non-graphical mode ? I thought it should not...

**Reply by HdeV999** [reply]
Well it can't, so it doesn't. I noticed my pipeline processing time got longer since my last repair of brainvisa (from 2 to 3 hours for 500 subjects) so I was looking at the logs and found this kind of logs which I didn't get previously. I also get a lot more bv_XXX files in my tmp folder than I used to, so I was wondering if deactivating the visualization could help: Here's a sample of logs which I find suspicious: 

Sample log 1:

Attribute Qt::AA_ShareOpenGLContexts must be set before QCoreApplication is created.
This plugin does not support propagateSizeHints()
cannot override an implicit transformation. Check for other paths in the transformation graph.
QOpenGLWidget: Failed to create context
warning: window geometry too small (dimx : 256, max: 0)
GLList::generate() could definitely not allocate OpenGL display list: no error


-----------------------------------------------------------

\##### ANATOMIST WARNING
AWindow3D::Refresh: OpenGL error.

-----------------------------------------------------------
-----------------------------------------------------------
\##### ANATOMIST ERROR
Light::RefreshGLList : OpenGL error in glGenLists.

-----------------------------------------------------------


Sample log 2:

\# Morphologist report

filtering...
after filtering: 1 components
Starting Anatomist.....
config file : /casa/home/.anatomist/config/settings.cfg
PyAnatomist Module present
PythonLauncher::runModules()
Load transformation : overriding existing one
Load transformation : overriding existing one
OpenGL lists cleanup...
No OpenGL extensions found
Number of texture units: 1
OpenGL lists cleanup...

Practically, my pipeline still runs, it's just longer and since I'm running it on many databases I was looking forward to revert it to its previous processing time, but I'm still happy with the current situation. 
It's probably unrelated, but I feel like I should highlight that in my brainvisa configuration, I had to change the [brainvisa]/src/axon/master/python/soma/qt4gui/text.py file to the one in https://brainvisa.info/download/updates/5.1.0/python/soma/qt4gui/text.py because of the specificities of the cluster I'm using...

**Reply by denisri** [reply]
So much longer, it's weird... The error or warning messages are "new" since we have replaced the way Anatomist renders offscreen, I have noticed similar messages but it seems to do the job anyway, and in my experiments this part has more or less the same performance as it used to, so I doubt this is the bottleneck. But yes an additional step has been added for the report. It's not that long (maybe 10 seconds per subject). We can try to unselect it just to see if it makes a difference. I'll send a modified process for Freesurfer importation when I have a minute, so that you can compare.

**Reply by HdeV999** [reply]
Thanks!

**Reply by denisri** [reply]
Hi,
Can you try this modified process ?
Save it in your `$HOME/.brainvisa/processes/` directory (if it doesn't exist, just create it) and restart brainvisa. It should replace the original process.
It has an additional parameter, `perform_morpho_report` (the last one) that you can un-check.
https://brainvisa.info/download/tmp/Import_FROM_FreeSurfer_TO_Morpho.py

**Reply by HdeV999** [reply]
Thanks a lot! Trying it with my current pipeline, will let you know as soon as it runs (my cluster queues can be a bit long).

**Reply by HdeV999** [reply]
It works! Here is an extract from my log files: 
progress: 91.66666666666666 % ...

Skip unselected node: morpho_report

progress: 100.0 % ...

  
 _Process **Morphologist 2023** finished on 2024/08/20 23:16 (14 minutes 59
seconds)_

OK

  
 _Process **Import FreeSurfer grey/white segmentation to Morphologist**
finished on 2024/08/20 23:16 (17 minutes 11 seconds)_

It didn't bring my processing time back to 2 hours though, but this may be related to other changes in my configuration. Thanks a lot!
